### ISSUE TITLE
2019-11-13 Extrapolated points not appearing in est_data output

### ISSUE THREAD

**Issue by jhhughes256** [tensfeldt/openNCA#7]
### Issue

When running a test case with a partial AUC including a point beyond the final observed value, the extrapolated point does not appear in the `est_data` output. Despite this, extrapolated points do seem to affect the value given in `data_out` for the partial AUC.

### Output

Using tc010 with two added partial AUCs from 0 to 120 and from 0 to 144. The first partial AUC includes extrapolation for some SDEIDs, while other SDEIDs have an observed value at this time. No SDEIDs have an observed value at 144.

The output from `est_data` is:

```
   PKDATAROWID    SDEID TIME  CEST_KEL   CEST_INT CEST_EXT CEST_C0 CEST_TLAST
1           NA  2925168    7        NA 3.59500000       NA      NA         NA
2     22289754  2925168   24 1.7026376         NA       NA      NA         NA
3     22289755  2925168   48 0.5878072         NA       NA      NA         NA
4           NA  2925168   64        NA 0.32866667       NA      NA         NA
5     22289756  2925168   72 0.2029306         NA       NA      NA  0.2029306
6           NA 40876220    7        NA 1.11575000       NA      NA         NA
7     22289571 40876220   16 0.4954443         NA       NA      NA         NA
8     22289572 40876220   24 0.3651747         NA       NA      NA         NA
9     22289573 40876220   48 0.1462239         NA       NA      NA         NA
10          NA 40876220   64        NA 0.08246667       NA      NA         NA
```

The extrapolated points do not appear in the `est_data`, while the interpolated points (from partial AUCs 0 to 7 and 0 to 64 from the original tc010) do appear.

The output from `data_out` for the same run is:

```
      SDEID     AUC0_7    AUC8_48     AUC0_64    AUC0_120    AUC0_144
1 111299859    5.15075   16.11586    23.59851    27.81464    29.81102
2 112146587   16.63043   27.18657    47.51955    50.43985    51.81112
3 123811839 1826.88593 1140.50993  3155.39210  3185.37920  3189.62671
4 145593586  759.77223 1482.00641  2543.66350  2692.11817  2708.33439
5 202367199 6447.26053 5524.24397 12915.26200 13167.20305 13196.88021
6 233891605 2026.37365 1432.19220  3683.11590  3729.02670  3733.35383
```

When comparing the partial AUCs given in `data_out` a difference between AUC0_120 and AUC0_144 is observed. This implies that extrapolation is likely occurring and being used to calculate the partial AUCs, but aren't being reported in `est_data`. This is consistent for all SDEIDs (including 2925168 shown in the `est_data` above).

### Reproducible Example

```
#### Load Libraries ------------------------------------------------------ ####
library(readr)
library(jsonlite)
library(testthat)
 
#### Load specific code-base for OpenNCA --------------------------------- ####
# pkgload::load_all(path = "codebase/openNCA-master/openNCA")
# pkgload::load_all(path = "codebase/openNCA-b1977923fc/openNCA")  # 11/8-v1
# pkgload::load_all(path = "codebase/openNCA-051805c887/openNCA")  # 11/8-v2
# pkgload::load_all(path = "codebase/openNCA-1fc479c2ca/openNCA")  # 11/11
# pkgload::load_all(path = "codebase/openNCA-a06c1c7461/openNCA")  # 11/12
pkgload::load_all(path = "codebase/openNCA-e3be44d927/openNCA")  # 11/13
 
#### Clean test environment ---------------------------------------------- ####
# Clean environment
rm(params, conc, vlist, results, results_list, r, e, mct, d, flags, p, klist, n, 
  ninc, nexc, k, rlist, dlist, nlist, slist, df, xmin, xmax, dk, flag_data, 
  est_data, outputdir)
 
profileid <- "SDEID"
args = commandArgs(trailingOnly=TRUE)
params = c()
 
#### Test Case 010 ------------------------------------------------------- ####
# M1 SD
# p <- "C:/Users/tensfeldt/Box Sync/EQuIP Computation Engine/Rudraya_executed_test_cases/tc010"
testcase.id <- "tc010"
p <- "testcase/tc010_M1SD"
params["INPUT1"] = file.path(p, "B1261002.in.csv")
params["parameterset"] ="PARAMETERLIST"
params["map"] = file.path(p, "tc010_mct_updated.csv")
params["flags"] = file.path(p, "B1261002_flag_data.in.csv")
params["OUTPUT"] ="."
print(params)
 
mct <- read_csv(params["map"])
write_file(toJSON(mct, na="string"), path=file.path(p, "tc010_mct_updated.json"))
params["map"] = file.path(p, "tc010_mct_updated.json")
 
flags <- read_csv(params["flags"])
write_file(toJSON(flags, na="string"), path=file.path(p, "tc010_flags.json"))
params["flags"] = file.path(p, "tc010_flags.json")
 
#### Read full concentration dataset ------------------------------------- ####
# Read conc
conc <- read.csv(file=params["INPUT1"], header=TRUE, sep=",", as.is=TRUE)
vlist <- c("PKDATAROWID", "PKPTMS", "PKCNCN", "PERIOD", "TREATXT")
 
# Read mct
mct <- params["map"]
###mct <- read_file(mct)
###mct <- gsub("[\\]+", "", mct)
mct <- fromJSON(mct)
 
# Read flags
flags <- params["flags"]
###flags <- read_file(flags)
###flags <- gsub("[\\]+", "", flags)
flags <- fromJSON(flags)
 
# ParameterSet
if(is.na(params["parameterset"])) { params["parameterset"] <- "PARAMETERDISPLAYLIST" }
parameterset <- params["parameterset"]
 
# FLAGS is a subset for the specific profile, subset concentration dataframe 
#  to those records by PKDATAROWID
d <- conc[is.element(conc$PKDATAROWID, flags$PKDATAROWID),]
 
# Add in missing required parameters for virtual compute call
klist <- c("KEL", "KELRSQ", "KELNOPT", "THALFF")
if(!is.element("PARAMETERDISPLAYLIST", names(mct))) { mct$PARAMETERDISPLAYLIST <- "" }
for(i in klist) {
  ###if(regexpr(i, mct$PARAMETERDISPLAYLIST, ignore.case=TRUE)==-1) { mct$PARAMETERDISPLAYLIST <- c(mct$PARAMETERDISPLAYLIST, i) }
  if(regexpr(i, mct$PARAMETERDISPLAYLIST, ignore.case=TRUE)==-1) { mct$PARAMETERDISPLAYLIST <- paste(mct$PARAMETERDISPLAYLIST, i, sep=";") }
}
 
# Remove any leading semi-colons ";"
mct$PARAMETERDISPLAYLIST <- gsub("^;*", "", mct$PARAMETERDISPLAYLIST, ignore.case=TRUE, perl=TRUE)
 
# For virtual compute call only: Subset the parameters to be calculated to those in parameterdisplaylist
if(!is.element("PARAMETERLIST", names(mct))) { mct$PARAMETERLIST <- "" }
 
# Check whether TIME selection has been set in MCT - default to NOMINAL
if(!is.element("TIME", names(mct))) { mct$TIME <- "NOMINAL" }
 
# Determine whether there are any points selected for inclusion in the terminal elimination phase via 
#  FLGEXKEL (0=not excluded/included; 1=excluded/not included)
#  ***What happens if there are no flags provided?***
n <- factor(flags$FLGEXKEL, levels=c(0,1))
ninc <- length(n[n==0])
nexc <- length(n[n==1])
#### Final changes to test ----------------------------------------------- ####
### Add partial AUC
mct$AUC.4.T1 <- 0
mct$AUC.4.T2 <- 120
mct$AUC.5.T1 <- 0
mct$AUC.5.T2 <- 144

#### Run computation engine ---------------------------------------------- ####
### Run computation
results_list <- run_computation(data = d, map = mct, flag = flags, 
  parameterset = parameterset)

### Save results to objects
r <- results_list$data_out
e <- results_list$est_data
f <- results_list$flag_data

### View output
head(e, 10)
e[e$TIME == 144,]  # no rows with time 144
head(r[c("SDEID", "AUC0_7", "AUC8_48", "AUC0_64", "AUC0_120", "AUC0_144")])
r$AUC0_144 - r$AUC0_120  # partial AUC 0-120 consistently smaller than 0-144
```

**Comment by jhhughes256** [comment]
### Update

As of 6284702d62 extrapolated points are now reported in `est_data` for example above. 

However, this issue still occurs for other test cases (it is uncertain if this error occurs for test cases currently affected by issue #6 #11 or #12  ). Additionally in these other test cases (tc012 and tc015), interpolated points aren't being reported either. 

Checking e3be44d927 this problem occurred in this previous commit as well.

### Output

When running test case 012 with an added partial AUC from 16 to 84 the following output is given:

```
   PKDATAROWID    SDEID TIME CEST_KEL CEST_INT CEST_EXT CEST_C0 CEST_TLAST
1     26908252 24280050   12 5.467556       NA       NA      NA         NA
2     26908253 24280050   18 2.696850       NA       NA      NA         NA
3     26908254 24280050   24 1.330210       NA       NA      NA   1.330210
4     26908343 45831434   12 6.053132       NA       NA      NA         NA
5     26908344 45831434   18 3.296653       NA       NA      NA         NA
6     26908345 45831434   24 1.795421       NA       NA      NA   1.795421
```

When running test case 015 with additional partial AUCs from 24 to 144 and 0 to 505 the following output is given:

```
   PKDATAROWID    SDEID TIME   CEST_KEL CEST_INT CEST_EXT CEST_C0 CEST_TLAST
1     33239236 12659123  168 1724.81135       NA       NA      NA         NA
2     33239237 12659123  336  663.38898       NA       NA      NA         NA
3     33239238 12659123  504  255.14961       NA       NA      NA  255.14961
4     33239594 39943270  168 2662.61782       NA       NA      NA         NA
5     33239595 39943270  336  838.84603       NA       NA      NA         NA
6     33239596 39943270  504  264.27475       NA       NA      NA  264.27475
7     33238557 61622147  120  145.89469       NA       NA      NA         NA
8     33238558 61622147  168   93.20112       NA       NA      NA         NA
```

As can be seen neither interpolated nor extrapolated points are showing up in est_data for tc012 and tc015.

### Reproducible Examples

#### tc012

```
#### Load Libraries ------------------------------------------------------ ####
library(readr)
library(jsonlite)
library(testthat)

#### Load specific code-base for OpenNCA --------------------------------- ####
# pkgload::load_all(path = "codebase/openNCA-master/openNCA")
# pkgload::load_all(path = "codebase/openNCA-a06c1c7461/openNCA")  # 11/12
# pkgload::load_all(path = "codebase/openNCA-e3be44d927/openNCA")  # 11/13
pkgload::load_all(path = "codebase/openNCA-6284702d62/openNCA")  # 11/14

#### Clean test environment ---------------------------------------------- ####
# Clean environment
rm(params, conc, vlist, results, results_list, r, e, mct, d, flags, p, klist, n, 
  ninc, nexc, k, rlist, dlist, nlist, slist, df, xmin, xmax, dk, flag_data, 
  est_data, outputdir)

profileid <- "SDEID"
args = commandArgs(trailingOnly=TRUE)
params = c()

#### Test Case 012 ------------------------------------------------------- ####
# M1 SS
# p <- "C:/Users/tensfeldt/Box Sync/EQuIP Computation Engine/Rudraya_executed_test_cases/tc012"
testcase.id <- "tc012"
p <- "testcase/tc012_M1SS"
params["INPUT1"] = file.path(p, "A0081216.in.csv")
params["parameterset"] ="PARAMETERLIST"
params["map"] = file.path(p, "tc012_mct_updated.csv")
params["flags"] = file.path(p, "A0081216_flag_data.in.csv")
params["OUTPUT"] ="."
print(params)

mct <- read_csv(params["map"])
write_file(toJSON(mct, na="string"), path=file.path(p, "tc012_mct_updated.json"))
params["map"] = file.path(p, "tc012_mct_updated.json")

flags <- read_csv(params["flags"])
write_file(toJSON(flags, na="string"), path=file.path(p, "tc012_flags.json"))
params["flags"] = file.path(p, "tc012_flags.json")

#### Read full concentration dataset ------------------------------------- ####
# Read conc
conc <- read.csv(file=params["INPUT1"], header=TRUE, sep=",", as.is=TRUE)
vlist <- c("PKDATAROWID", "PKPTMS", "PKCNCN", "PERIOD", "TREATXT")

# Read mct
mct <- params["map"]
###mct <- read_file(mct)
###mct <- gsub("[\\]+", "", mct)
mct <- fromJSON(mct)

# Read flags
flags <- params["flags"]
###flags <- read_file(flags)
###flags <- gsub("[\\]+", "", flags)
flags <- fromJSON(flags)

# ParameterSet
if(is.na(params["parameterset"])) { params["parameterset"] <- "PARAMETERDISPLAYLIST" }
parameterset <- params["parameterset"]

# FLAGS is a subset for the specific profile, subset concentration dataframe 
#  to those records by PKDATAROWID
d <- conc[is.element(conc$PKDATAROWID, flags$PKDATAROWID),]

# Add in missing required parameters for virtual compute call
klist <- c("KEL", "KELRSQ", "KELNOPT", "THALFF")
if(!is.element("PARAMETERDISPLAYLIST", names(mct))) { mct$PARAMETERDISPLAYLIST <- "" }
for(i in klist) {
  ###if(regexpr(i, mct$PARAMETERDISPLAYLIST, ignore.case=TRUE)==-1) { mct$PARAMETERDISPLAYLIST <- c(mct$PARAMETERDISPLAYLIST, i) }
  if(regexpr(i, mct$PARAMETERDISPLAYLIST, ignore.case=TRUE)==-1) { mct$PARAMETERDISPLAYLIST <- paste(mct$PARAMETERDISPLAYLIST, i, sep=";") }
}

# Remove any leading semi-colons ";"
mct$PARAMETERDISPLAYLIST <- gsub("^;*", "", mct$PARAMETERDISPLAYLIST, ignore.case=TRUE, perl=TRUE)

# For virtual compute call only: Subset the parameters to be calculated to those in parameterdisplaylist
if(!is.element("PARAMETERLIST", names(mct))) { mct$PARAMETERLIST <- "" }

# Check whether TIME selection has been set in MCT - default to NOMINAL
if(!is.element("TIME", names(mct))) { mct$TIME <- "NOMINAL" }

# Determine whether there are any points selected for inclusion in the terminal elimination phase via 
#  FLGEXKEL (0=not excluded/included; 1=excluded/not included)
#  ***What happens if there are no flags provided?***
n <- factor(flags$FLGEXKEL, levels=c(0,1))
ninc <- length(n[n==0])
nexc <- length(n[n==1])

#### Final changes to test ----------------------------------------------- ####
### Add partial AUC
mct$AUC.1.T1 <- 16
mct$AUC.1.T2 <- 84
# mct$AUC.2.T1 <- 0
# mct$AUC.2.T2 <- 409

#### Run computation engine ---------------------------------------------- ####
### Run computation
# pretime <- Sys.time()
results_list <- run_computation(data = d, map = mct, flag = flags, 
  parameterset = parameterset)
# Sys.time() - pretime

### Save results to objects
r <- results_list$data_out
e <- results_list$est_data
f <- results_list$flag_data

### Show results
head(e, 10)
head(r[c("SDEID", "AUC16_84")])
```

#### tc015

```
#### Load Libraries ------------------------------------------------------ ####
library(readr)
library(jsonlite)
library(testthat)

#### Load specific code-base for OpenNCA --------------------------------- ####
# pkgload::load_all(path = "codebase/openNCA-master/openNCA")
# pkgload::load_all(path = "codebase/openNCA-a06c1c7461/openNCA")  # 11/12
# pkgload::load_all(path = "codebase/openNCA-e3be44d927/openNCA")  # 11/13
pkgload::load_all(path = "codebase/openNCA-6284702d62/openNCA")  # 11/14

#### Clean test environment ---------------------------------------------- ####
# Clean environment
rm(params, conc, vlist, results, results_list, r, e, mct, d, flags, p, klist, n, 
  ninc, nexc, k, rlist, dlist, nlist, slist, df, xmin, xmax, dk, flag_data, 
  est_data, outputdir)

profileid <- "SDEID"
args = commandArgs(trailingOnly=TRUE)
params = c()

#### Test Case 015 ------------------------------------------------------- ####
# M3 SD
# p <- "C:/Users/tensfeldt/Box Sync/EQuIP Computation Engine/Rudraya_executed_test_cases/tc015"
testcase.id <- "tc015"
p <- "testcase/tc015_M3SD"
params["INPUT1"] = file.path(p, "B2901001.in.csv")
params["parameterset"] ="PARAMETERLIST"
params["map"] = file.path(p, "tc015_mct_updated.csv")
params["flags"] = file.path(p, "B2901001_flag_data.in.csv")
params["OUTPUT"] ="."
print(params)

mct <- read_csv(params["map"])
write_file(toJSON(mct, na="string"), path=file.path(p, "tc015_mct_updated.json"))
params["map"] = file.path(p, "tc015_mct_updated.json")

flags <- read_csv(params["flags"])
write_file(toJSON(flags, na="string"), path=file.path(p, "tc015_flags.json"))
params["flags"] = file.path(p, "tc015_flags.json")

#### Read full concentration dataset ------------------------------------- ####
# Read conc
conc <- read.csv(file=params["INPUT1"], header=TRUE, sep=",", as.is=TRUE)
vlist <- c("PKDATAROWID", "PKPTMS", "PKCNCN", "PERIOD", "TREATXT")

# Read mct
mct <- params["map"]
###mct <- read_file(mct)
###mct <- gsub("[\\]+", "", mct)
mct <- fromJSON(mct)

# Read flags
flags <- params["flags"]
###flags <- read_file(flags)
###flags <- gsub("[\\]+", "", flags)
flags <- fromJSON(flags)

# ParameterSet
if(is.na(params["parameterset"])) { params["parameterset"] <- "PARAMETERDISPLAYLIST" }
parameterset <- params["parameterset"]

# FLAGS is a subset for the specific profile, subset concentration dataframe 
#  to those records by PKDATAROWID
d <- conc[is.element(conc$PKDATAROWID, flags$PKDATAROWID),]

# Add in missing required parameters for virtual compute call
klist <- c("KEL", "KELRSQ", "KELNOPT", "THALFF")
if(!is.element("PARAMETERDISPLAYLIST", names(mct))) { mct$PARAMETERDISPLAYLIST <- "" }
for(i in klist) {
  ###if(regexpr(i, mct$PARAMETERDISPLAYLIST, ignore.case=TRUE)==-1) { mct$PARAMETERDISPLAYLIST <- c(mct$PARAMETERDISPLAYLIST, i) }
  if(regexpr(i, mct$PARAMETERDISPLAYLIST, ignore.case=TRUE)==-1) { mct$PARAMETERDISPLAYLIST <- paste(mct$PARAMETERDISPLAYLIST, i, sep=";") }
}

# Remove any leading semi-colons ";"
mct$PARAMETERDISPLAYLIST <- gsub("^;*", "", mct$PARAMETERDISPLAYLIST, ignore.case=TRUE, perl=TRUE)

# For virtual compute call only: Subset the parameters to be calculated to those in parameterdisplaylist
if(!is.element("PARAMETERLIST", names(mct))) { mct$PARAMETERLIST <- "" }

# Check whether TIME selection has been set in MCT - default to NOMINAL
if(!is.element("TIME", names(mct))) { mct$TIME <- "NOMINAL" }

# Determine whether there are any points selected for inclusion in the terminal elimination phase via 
#  FLGEXKEL (0=not excluded/included; 1=excluded/not included)
#  ***What happens if there are no flags provided?***
n <- factor(flags$FLGEXKEL, levels=c(0,1))
ninc <- length(n[n==0])
nexc <- length(n[n==1])

#### Final changes to test ----------------------------------------------- ####
### Add partial AUC
mct$AUC.2.T1 <- 24
mct$AUC.2.T2 <- 144
mct$AUC.3.T1 <- 0
mct$AUC.3.T2 <- 505

#### Run computation engine ---------------------------------------------- ####
### Run computation
# pretime <- Sys.time()
results_list <- run_computation(data = d, map = mct, flag = flags, 
  parameterset = parameterset)

### Save results to objects
r <- results_list$data_out
e <- results_list$est_data
f <- results_list$flag_data

### Show results
head(e, 10)
head(r[c("SDEID", "AUC0_120", "AUC24_144", "AUC0_505")])
```

**Comment by jhhughes256** [comment]
### Update

This behaviour is still seen in the following test cases for commit ae10cfa1d9:

- tc11 (run with partial AUC 0-x to avoid 'subscript out of bounds')
- tc12 (run with partial AUC 0-x to avoid 'subscript out of bounds')
- tc13 (run with partial AUC 0-x to avoid 'subscript out of bounds')
- tc14 (run with partial AUC 0-x to avoid 'subscript out of bounds')
- tc15 (run with partial AUC 0-x to avoid 'subscript out of bounds')

**Comment by jhhughes256** [comment]
This issue for tc011 was mentioned in #21

**Comment by jhhughes256** [comment]
### Update

This issue has been fixed as of 066fa5d8a1.
